### PR TITLE
Display fatal error and delete session if sql context creation errors…

### PR DESCRIPTION
**Pull request by MohamedElKamhawy**
Not tested yet, need to figure out a way to repro the way I found the error

### CODE REVIEW

**Comment by msftristew** [comment]
If you remove the `if` statement and just raise a `ValueError` at that location, you'll be able to see what the output would look like from the notebook view.

**Comment by aggFTW** [comment]
And you should have unit tests for this, from the magics all the way to LivySession

**Comment by aggFTW** [comment]
I'll review once unit tests have been added all the way through.

**Comment by MohamedElKamhawy** [comment]
I tested the code locally and added the needed unit tests.

**Comment by msftristew** [comment]
1. Is the error message the user gets sensible?

2. Does it prevent the user from executing more cells when this fails?

**Comment by MohamedElKamhawy** [comment]
@msftristew  I mentioned "Failed to create the SqlContext" and added the exception. 
If the user tries again to run any cell, he'll get the same message. (above)

**Comment by MohamedElKamhawy** [comment]
It seems the build fails.

**Comment by msftristew** [comment]
merge when tests pass

**Comment by msftristew** [comment]
And when Alejandro says OK :)

File: tests/test_livysession.py [hunk 1]
```diff
@@ -436,6 +437,21 @@ def test_create_sql_hive_context_happens_once(self):
...
+
+        session._get_sql_context_creation_command.return_value.execute.return_value = (True, "Success")
+        session.created_sql_context = None
+        session.create_sql_context()
+        assert session.created_sql_context
+
+        session._get_sql_context_creation_command.return_value.execute.return_value = (False, "Exception")
+        session.created_sql_context = None
+        try:
+            session.create_sql_context()
```
Comment (msftristew) [inline tests/test_livysession.py]: You need to insert another `assert` in here to make sure that the function threw an exception

Comment (MohamedElKamhawy) [inline tests/test_livysession.py]: I added it now in a new commit

File: tests/test_livysession.py [hunk 2]
```diff
@@ -39,7 +39,8 @@ class TestLivySession:
     ready_statement_json = json.loads('{"id":0,"state":"available","output":{"status":"ok","execution_count":0,"data":{"text/plain":"Pi is roughly 3.14336"}}}')
     log_json = json.loads('{"id":6,"from":0,"total":212,"log":["hi","hi"]}')
 
-    def __init__(self):
+    def __init__(self, methodName="test_create_sql_hive_context_happens_once"):
```
Comment (msftristew) [inline tests/test_livysession.py]: What does this do?

Comment (MohamedElKamhawy) [inline tests/test_livysession.py]: This allows me to make sure this method is throwing an exception.

Comment (msftristew) [inline tests/test_livysession.py]: I am asking specifically what this `methodName` addition is for.

Comment (MohamedElKamhawy) [inline tests/test_livysession.py]: I'm inheriting unittest class and this is the way to call the super class by specifying the method that you apply your test case to.

Comment (msftristew) [inline tests/test_livysession.py]: But the `TestLivySession` class is way bigger than just your test.

File: tests/test_livysession.py (hunk 1 above)
Comment (msftristew) [inline tests/test_livysession.py]: The change you made wasn't really necessary, just `assert False` at this location (like I suggested) to make sure an exception was thrown

File: tests/test_livysession.py (hunk 2 above)
Comment (MohamedElKamhawy) [inline tests/test_livysession.py]: Talked face to face, all tests are running and the current work is okay.

**Comment by MohamedElKamhawy** [comment]
@aggFTW  Can you have a look?

File: remotespark/livyclientlib/livysession.py [hunk 3]
```diff
@@ -90,11 +89,14 @@ def create_sql_context(self):
         self.ipython_display.writeln("Creating HiveContext as 'sqlContext'")
         command = self._get_sql_context_creation_command()
         try:
-            command.execute(self)
+            (success, out) = command.execute(self)
         except LivyClientTimeoutError:
             raise LivyClientTimeoutError("Failed to create the SqlContext in time. Timed out after {} seconds."
                                          .format(self._wait_for_idle_timeout_seconds))
-        self.created_sql_context = True
+        if success:
```
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: You just modified the behavior of livysession (it can now raise a ValueError). What class consumes this exception? I think it should be the magics and should set a fatal error as True. Is that already happening here?

For this to hold True, all intermediate classes between livysession and kernelmagics must not catch any ValueErrors in the methods that get called. Is that already True? Can we unit test that?

File: remotespark/livyclientlib/livysession.py [hunk 4]
```diff
@@ -90,11 +89,14 @@ def create_sql_context(self):
...
+            self.created_sql_context = True
+        else:
+            raise ValueError("Failed to create the SqlContext.\nError {}".format(out))
```
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: I would put a semicolon after Error and quote the actual error:
`raise ValueError("Failed to create the SqlContext.\nError: '{}'".format(out))`

What do you think?

File: tests/test_livysession.py [hunk 5]
```diff
@@ -436,6 +436,22 @@ def test_create_sql_hive_context_happens_once(self):
...
+
+        session._get_sql_context_creation_command.return_value.execute.return_value = (False, "Exception")
+        session.created_sql_context = None
+
+        try:
+            session.create_sql_context()
+            assert False
+        except ValueError:
+            assert session.created_sql_context is None
+
```
Comment (aggFTW) [inline tests/test_livysession.py]: This assertion was valuable:

`assert ex.message == "Failed to create the SqlContext.\nError {}".format("Exception")`

Can you please add it back?

File: remotespark/livyclientlib/livysession.py (hunk 3 above)
Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: Yes, this is true, no class catches this exception at all.
I already tested this and none catches this exception.

File: tests/test_livysession.py (hunk 5 above)
Comment (MohamedElKamhawy) [inline tests/test_livysession.py]: message attribute is not existing in Python 3 and I believe the current test is sufficient.

File: remotespark/livyclientlib/livysession.py (hunk 4 above)
Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: As a user, I prefer to see the summary first and in the second line to see the root cause for the error.

Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: But, I'm fine with either.

Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: What I suggested keeps the number of lines and placement. It's just adding a semi-colon and quotes to denote that the Error is not coming from us, but from Spark.

File: tests/test_livysession.py (hunk 5 above)
Comment (aggFTW) [inline tests/test_livysession.py]: You can always get the ValueError text. Just do a str(e).

File: remotespark/livyclientlib/livysession.py (hunk 3 above)
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: It is true now, but without the unit tests to all those intermediate classes, you won't know that it will never change. Once you write those tests, it would be good to say why those classes should not catch ValueErrors in the unit tests.

Please add them to prevent regressions.

File: remotespark/livyclientlib/livysession.py (hunk 4 above)
Comment (msftristew) [inline remotespark/livyclientlib/livysession.py]: That's a colon, not a semicolon :)

Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: Thanks Ricky :)

File: remotespark/livyclientlib/livysession.py (hunk 3 above)
Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: Can you please explain more about what do you think about that?

Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: Start method - in LivySession Class - throw an exception if the session is not created (Id < 0), I can't find the relevant unit test for that? Can you point me to it to do the same?

Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: Any help?

Comment (msftristew) [inline remotespark/livyclientlib/livysession.py]: I'm advising Mohamed to add, at the very least, to add a unit test to the magics class to ensure it logs a fatal error when adding the session fails (if that unit test doesn't already exist).  He can also add a unit test to `SparkController` to verify it raises an exception if `session.start()` throws which would be very easy.  That would be enough verification.

Another thing you can do for some further assurance that this exception will not be spontaneously caught is to use a different exception than `ValueError`, like a custom exception `FailedToCreateSqlContextError`.

**Comment by MohamedElKamhawy** [comment]
Addressing Feedback & Added the unit tests for the intermediate classes.

File: remotespark/livyclientlib/livysession.py (hunk 3 above)
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: This feedback is not addressed on the latest iteration. We meant that you should at least add a unit test to the test_sparkcontroller.py file for when session.start() fails as well as a unit test to the test_kernelmagics.py file for when sparkcontroller.add_session() fails.

File: tests/test_livysession.py [hunk 6]
```diff
@@ -436,6 +438,49 @@ def test_create_sql_hive_context_happens_once(self):
                in http_client.post_statement.call_args_list
         assert len(http_client.post_statement.call_args_list) == 1
 
+        session._get_sql_context_creation_command = MagicMock()
```
Comment (aggFTW) [inline tests/test_livysession.py]: This test is massive now. Can you please break it up into different tests. The current test should only `test_create_sql_hive_context_happens_once`. This test is now testing what happens on failure.

File: tests/test_livysession.py [hunk 7]
```diff
@@ -26,7 +27,7 @@ def status_code(self):
 CONN_STR = 'url=https://www.DFAS90D82309F0W9ASD0F9ZX.com;username=abcd;password=1234'
 
 
-class TestLivySession:
+class TestLivySession():
```
Comment (msftristew) [inline tests/test_livysession.py]: `TestLivySession` should inherit from `object`

Comment (MohamedElKamhawy) [inline tests/test_livysession.py]: This is not part of my PR, I'll add it now.

File: tests/test_livysession.py (hunk 6 above)
Comment (MohamedElKamhawy) [inline tests/test_livysession.py]: Agreed

**Comment by MohamedElKamhawy** [comment]
Addressing the feedback.

File: tests/test_livysession.py [hunk 8]
```diff
@@ -436,6 +437,72 @@ def test_create_sql_hive_context_happens_once(self):
...
+            "status_sleep_seconds": 0.01,
+            "statement_sleep_seconds": 0.01
+        })
+        session = self._create_session(kind=kind, http_client=http_client)
+        session.ipython_display = ipython_display
+        conf.load()
+
+        session._get_sql_context_creation_command = MagicMock()
+
+        session._get_sql_context_creation_command.return_value.execute.return_value = (True, "Success")
```
Comment (aggFTW) [inline tests/test_livysession.py]: This is testing success, when the unit test name is `test_create_sql_context_failure`. Can you please break it up into different unit tests?

Also, the setup code is copy pasted from another unit test. It warrants a helper function.

File: tests/test_livysession.py [hunk 9]
```diff
@@ -436,6 +437,72 @@ def test_create_sql_hive_context_happens_once(self):
...
+        session.created_sql_context = None
+
+        try:
+            session.create_sql_context()
+            assert False
+        except ValueError as ex:
+            assert str(ex) == "Failed to create the SqlContext.\nError, '{}'".format("Exception")
+            assert session.created_sql_context is None
+
+        kernelMagics = KernelMagics(shell=None)
```
Comment (msftristew) [inline tests/test_livysession.py]: Why is this here and not in `test_kernel_magics` (the file where we're doing the rest of the testing of kernel magics)?

Comment (aggFTW) [inline tests/test_livysession.py]: What is a kernel magics test doing in test_livysession.py?

Why is this unit test testing more than one thing again?

File: tests/test_livysession.py [hunk 10]
```diff
@@ -436,6 +437,72 @@ def test_create_sql_hive_context_happens_once(self):
...
+        kernelMagics.language = constants.LANG_SCALA
+        kernelMagics.ipython_display = ipython_display
+        kernelMagics.ipython_display.send_error = MagicMock()
+
+        kernelMagics._do_not_call_start_session("Test Line")
+        assert kernelMagics.fatal_error
+        assert kernelMagics.fatal_error_message == conf.fatal_error_suggestion().format(
+            "Failed to create the SqlContext.\nError, '{}'".format("Exception"))
+
+        sparkController = SparkController(ipython_display)
```
Comment (aggFTW) [inline tests/test_livysession.py]: What is a spark controller test doing in test_livysession.py?

Why is this unit test testing more than one thing again?

**Comment by aggFTW** [comment]
This is very important:

1. Unit tests should only be testing one thing at a time. **Maybe** two if the two things are very similar... but they should be testing a **unit*. This is so that when a unit test fails, we know exactly what is failing. If the unit test is testing 4 things at once and it fails, which of the 4 things failed? I need to go debug to know that and that is wasted time.
2. Every class is in a different file. A class is some code that is isolated of its dependencies. So, if you are _unit testing_ a class, you should be testing it isolated of its dependencies, so that even if dependencies fail, your unit tests do not get affected. Unit tests are grouped into different files because every test file unit tests just one class. If you are importing dependencies, [like this](https://github.com/jupyter-incubator/sparkmagic/pull/198/files#diff-498e0d5682020c4fc81dfb940efe63feR8), it's a pretty good indication of a [code smell](https://en.wikipedia.org/wiki/Code_smell).

So, please break your unit tests into **unit** tests and put them in the files they belong at.

File: remotespark/livyclientlib/livysession.py (hunk 3 above)
Comment (MohamedElKamhawy) [inline remotespark/livyclientlib/livysession.py]: Done

**Comment by MohamedElKamhawy** [comment]
I have pushed a new one, and applied the different feedback I got. 
@msftristew @aggFTW  Thank you for your patience, and for helping. :)

File: tests/test_sparkcontroller.py [hunk 11]
```diff
@@ -193,3 +193,24 @@ def test_get_logs():
...
+@with_setup(_setup, _teardown)
+def test_start_throw_exception():
+    name = "name"
+    properties = {"kind": "spark"}
+    endpoint = Endpoint("http://location:port", "name", "word")
+    session = MagicMock()
+
+    controller._livy_session = MagicMock(return_value=session)
+    controller._http_client = MagicMock(return_value=MagicMock())
+    session.start = dummy_create_sql_context
```
Comment (msftristew) [inline tests/test_sparkcontroller.py]: Use `MagicMock` for this, then you can also assert the session was started

File: tests/test_kernel_magics.py [hunk 12]
```diff
@@ -417,3 +417,20 @@ def test_delete_with_force_different_session():
...
+    magic.ipython_display = ipython_display
+    magic.ipython_display.send_error = MagicMock()
+
+    magic._do_not_call_start_session("Test Line")
+    assert magic.fatal_error
+    assert magic.fatal_error_message == conf.fatal_error_suggestion().format(
+        "Failed to create the SqlContext.\nError, '{}'".format("Exception"))
+
+
+def dummy_add_session(*args):
```
Comment (msftristew) [inline tests/test_kernel_magics.py]: Use MagicMock for this, then you can also assert the method was called

Comment (MohamedElKamhawy) [inline tests/test_kernel_magics.py]: Added in the new push

File: tests/test_sparkcontroller.py (hunk 11 above)
Comment (MohamedElKamhawy) [inline tests/test_sparkcontroller.py]: Added in the new push

**Comment by msftristew** [comment]
LGTM, just resolve your conflicts

File: tests/test_kernel_magics.py [hunk 13]
```diff
@@ -417,3 +417,20 @@ def test_delete_with_force_different_session():
 
     spark_controller.get_session_id_for_client.assert_called_once_with(magic.session_name)
     spark_controller.delete_session_by_id.assert_called_once_with(magic.endpoint, session_id)
+
+
+@with_setup(_setup, _teardown)
+def test_add_session_throw_exception():
```
Comment (aggFTW) [inline tests/test_kernel_magics.py]: A better name would be `test_start_session_displays_fatal_error_when_session_throws`

Comment (MohamedElKamhawy) [inline tests/test_kernel_magics.py]: isn't it too big?

File: tests/test_livysession.py [hunk 14]
```diff
@@ -439,6 +440,32 @@ def test_create_sql_hive_context_happens_once(self):
...
         assert len(http_client.post_statement.call_args_list) == 1
 
+        session._get_sql_context_creation_command = MagicMock()
+
+        session._get_sql_context_creation_command.return_value.execute.return_value = (True, "Success")
+        session.created_sql_context = None
+        session.create_sql_context()
+        assert session.created_sql_context
+
+    def test_create_sql_context_throw_exception(self):
```
Comment (aggFTW) [inline tests/test_livysession.py]: `test_create_sql_context_throws_when_command_fails`

File: tests/test_kernel_magics.py (hunk 13 above)
Comment (msftristew) [inline tests/test_kernel_magics.py]: No

File: tests/test_sparkcontroller.py [hunk 15]
```diff
@@ -193,3 +193,23 @@ def test_get_logs():
 def test_get_session_id_for_client():
     assert controller.get_session_id_for_client("name") is not None
     client_manager.get_session_id_for_client.assert_called_once_with("name")
+
+
+@with_setup(_setup, _teardown)
+def test_start_throw_exception():
```
Comment (aggFTW) [inline tests/test_sparkcontroller.py]: `test_add_session_throws_when_session_start_fails`

Comment (MohamedElKamhawy) [inline tests/test_sparkcontroller.py]: Any other comments? If no, will update per merging.

File: tests/test_sparkcontroller.py [hunk 16]
```diff
@@ -193,3 +193,23 @@ def test_get_logs():
...
+    controller._livy_session = MagicMock(return_value=session)
+    controller._http_client = MagicMock(return_value=MagicMock())
+    e = ValueError("Failed to create the SqlContext.\nError, '{}'".format("Exception"))
+    session.start = MagicMock(side_effect=e)
+
+    try:
+        controller.add_session(name, endpoint, False, properties)
+        assert False
+    except ValueError as ex:
+        assert str(ex) == "Failed to create the SqlContext.\nError, '{}'".format("Exception")
```
Comment (aggFTW) [inline tests/test_sparkcontroller.py]: `assert str(ex) == str(e)`

File: tests/test_kernel_magics.py (hunk 13 above)
Comment (aggFTW) [inline tests/test_kernel_magics.py]: Too big for?

Comment (MohamedElKamhawy) [inline tests/test_kernel_magics.py]: As a method name, it's too big.

Comment (msftristew) [inline tests/test_kernel_magics.py]: No

Comment (aggFTW) [inline tests/test_kernel_magics.py]: The method name is too big for what?

**Comment by aggFTW** [comment]
Mohamed, please don't merge until we see the conflicts resolved. So let us know when you resolve the conflicts and we'll take a last look.

File: tests/test_kernel_magics.py (hunk 13 above)
Comment (MohamedElKamhawy) [inline tests/test_kernel_magics.py]: Too big from naming convention perspective :)

Comment (aggFTW) [inline tests/test_kernel_magics.py]: What naming convention?

I prefer clarity over anything? What about you?

Comment (MohamedElKamhawy) [inline tests/test_kernel_magics.py]: OK

**Comment by MohamedElKamhawy** [comment]
Please, have a look to merge.

File: tests/test_livysession.py [hunk 17]
```diff
@@ -1,11 +1,12 @@
 ﻿import json
-
 from mock import MagicMock, call
 from nose.tools import raises, assert_equals
 
 from remotespark.livyclientlib.livyclienttimeouterror import LivyClientTimeoutError
 from remotespark.livyclientlib.livyunexpectedstatuserror import LivyUnexpectedStatusError
 from remotespark.livyclientlib.livysession import LivySession
+from remotespark.kernels.kernelmagics import KernelMagics
+from remotespark.livyclientlib.sparkcontroller import SparkController
```
Comment (aggFTW) [inline tests/test_livysession.py]: These imports are not needed

Comment (MohamedElKamhawy) [inline tests/test_livysession.py]: Sure, any other comments?

File: tests/test_kernel_magics.py [hunk 18]
```diff
@@ -531,6 +531,21 @@ def test_delete_with_force_different_session():
...
+    magic.ipython_display = ipython_display
+    magic.ipython_display.send_error = MagicMock()
+
+    magic._do_not_call_start_session("Test Line")
+
+    magic.spark_controller.add_session.assert_called_once_with(magic.session_name, magic.endpoint, False,
+                                                               {"kind": constants.SESSION_KIND_SPARK})
+    assert magic.fatal_error
+    assert magic.fatal_error_message == conf.fatal_error_suggestion().format(str(e))
+
```
Comment (aggFTW) [inline tests/test_kernel_magics.py]: Two new lines between tests.

**Comment by aggFTW** [comment]
Looks good! Just the two small comments :smile:

**Comment by aggFTW** [comment]
Nice! Merge after Travis passes.

**Comment by MohamedElKamhawy** [comment]
closes #198
